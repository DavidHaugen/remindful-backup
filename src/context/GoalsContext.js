import React, { Component } from 'react'


const GoalsContext = React.createContext({
  goals: [],
  error: null,
  setError: () => {},
  clearError: () => {},
  setGoals: () => {}
})

export default GoalsContext

export class GoalsProvider extends Component {
  state = {
    goals: [],
    error: null,
  };

  setGoals = goals => {
    this.setState({ goals })
  }

  setError = error => {
    console.error(error)
    this.setState({ error })
  }

  clearError = () => {
    this.setState({ error: null })
  }

  render() {
    const value = {
      goals: this.state.goals,
      error: this.state.error,
      setError: this.setError,
      clearError: this.clearError,
      setGoals: this.setGoals
    }

    return (
      <GoalsContext.Provider value={value}>
        {this.props.children}
      </GoalsContext.Provider>
    )
  }
}