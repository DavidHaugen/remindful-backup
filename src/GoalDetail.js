import React, { Component } from 'react'
import GoalsContext from './context/GoalsContext'
import remindfulApiService from './services/remindful-api-service';

class GoalDetail extends Component {

  static contextType = GoalsContext

  state = {
    isEditable: false,
    reflections: []
  }

  submitReflection = e => {
    e.preventDefault();
    const {reflections} = e.target
    const goalId = this.props.match.params.goalId
    remindfulApiService.postNewReflection(goalId, reflections.value)
    .then(reflection => this.setState({reflections: [...this.state.reflections, reflection]}))
    .then(document.getElementById('reflectionsForm').reset())
  }

  deleteGoal = goal => {
    remindfulApiService.deleteGoal(goal);
    this.context.deleteGoal(goal);
    this.props.history.push('/my-goals');
  }

  componentDidMount (){
    remindfulApiService.getReflections(Number(this.props.match.params.goalId))
      .then(reflections => this.setState({reflections}))
  }
  

  render(){
    const reflections = this.state.reflections.map((reflection, key) =>{ 
      const date = new Date(reflection.date_created)
      return <li key={key}>
          <span className='reflectionDate'>Date: {(date.getMonth() + 1) + '/' + date.getDate() + '/' +  date.getFullYear()}</span><span className='reflectionContent'>{reflection.content}</span><button id={reflection.id}>delete</button>
        </li>
    }
    )
    let [goal] = this.context.goals.filter((goal) => (goal.id === Number(this.props.match.params.goalId)))
    return (
      <div>
        <h1>Your goal:</h1>
        <h2>{
          (goal) ? 
          goal.name : null}</h2>
        <h3>Reflections on this goal:</h3>
        <ul>
          {reflections}
        </ul>
          <form id='reflectionsForm' onSubmit={(e)=>this.submitReflection(e)}>
            <input id='reflections' name='reflections' >
                </input>
            <button type='submit' >Submit reflection</button>
          </form>
        <button onClick={() => this.deleteGoal(this.props.match.params.goalId)}>Delete goal</button>
        <button >Mark complete</button>

      </div>
    )
  }
}

export default GoalDetail