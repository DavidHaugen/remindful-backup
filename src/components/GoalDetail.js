import React, { Component } from 'react'
import GoalsContext from '../context/GoalsContext'
import remindfulApiService from '../services/remindful-api-service';
import '../styles/GoalDetails.css'

class GoalDetail extends Component {

  static contextType = GoalsContext

  state = {
    isEditable: false,
    reflections: []
  }

  submitReflection = e => {
    e.preventDefault();
    const {reflections} = e.target
    const goalId = this.props.match.params.goalId
    remindfulApiService.postNewReflection(goalId, reflections.value)
    .then(reflection => this.setState({reflections: [...this.state.reflections, reflection]}))
    .then(document.getElementById('reflectionsForm').reset())
  }

  deleteGoal = goal => {
    remindfulApiService.deleteGoal(goal);
    this.context.deleteGoal(goal);
    this.props.history.push('/my-goals');
  }

  deleteReflection = (e) => {
    e.preventDefault();
    const id = e.target.id;
    remindfulApiService.deleteReflection(id)
    console.log(id)
    const targetReflection = this.state.reflections.filter((reflection) => reflection.id === Number(id));
    const newReflections = this.state.reflections;
    newReflections.splice(this.state.reflections.indexOf(targetReflection[0]), 1);
    this.setState({reflections: newReflections})
  }

  markGoalComplete = id => {
    const targetGoal = this.context.goals.filter((goal) => goal.id === Number(id));
    const updatedGoal = {...targetGoal[0]}
    updatedGoal.complete = !updatedGoal.complete
    remindfulApiService.markGoalComplete(updatedGoal)
    this.context.markGoalComplete(id)
  }

  componentDidMount (){
    remindfulApiService.getReflections(Number(this.props.match.params.goalId))
      .then(reflections => this.setState({reflections}))
  }
  

  render(){
    const reflections = this.state.reflections.map((reflection, key) =>{ 
      const date = new Date(reflection.date_created)
      return <li key={key}>
          <span className='reflectionDate'>Date: {(date.getMonth() + 1) + '/' + date.getDate() + '/' +  date.getFullYear()}</span><span className='reflectionContent'>{reflection.content}</span><button id={reflection.id} onClick={(e) => this.deleteReflection(e)}>delete</button>
        </li>
    }
    )
    let [goal] = this.context.goals.filter((goal) => (goal.id === Number(this.props.match.params.goalId)))
    return (
      <div>
        <h1>Your goal:</h1>
        <h2 className={
          (goal) ? 
          goal.complete ? 'complete' : 'inProgress' :
        null}>{
          (goal) ? 
          goal.name : null}</h2>
        <h3>Reflections on this goal:</h3>
        <ul>
          {reflections}
        </ul>
          <form id='reflectionsForm' onSubmit={(e)=>this.submitReflection(e)}>
            <input id='reflections' name='reflections' >
                </input>
            <button type='submit' >Submit reflection</button>
          </form>
        <button onClick={() => this.deleteGoal(this.props.match.params.goalId)}>Delete goal</button>
        <button onClick={() => this.markGoalComplete(this.props.match.params.goalId)}>Mark complete</button>

      </div>
    )
  }
}

export default GoalDetail