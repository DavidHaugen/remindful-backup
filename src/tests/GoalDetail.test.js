import React from 'react';
import {MemoryRouter} from 'react-router-dom'
import GoalDetail from '../components/GoalDetail'
import {mount} from 'enzyme';
import GoalsContext from '../context/GoalsContext';
import renderer from 'react-test-renderer'

describe('<GoalDetail />', () => {
  it('Renders without crashing', () => {
    const state ={ goals: [{complete: false, id: 1, name: 'test'}, {complete: false, id: 2, name: 'test2'}],
    loadingTrue: () => {},
    loadingFalse: () => {}
  }
      mount(<MemoryRouter><GoalsContext.Provider value={state}><GoalDetail match={
        {params: {goalId: 1}}
      }/></GoalsContext.Provider></MemoryRouter>);
  });

  // it("renders without crashing", () => {
  //   const div = document.createElement("div");
  //   ReactDOM.render(
  //     <MemoryRouter>
  //       <GoalDetail />
  //     </MemoryRouter>,
  //     div
  //   ); 
  // })

  it('renders the UI as expected', () => {
    const state ={ goals: [{complete: false, id: 1, name: 'test'}, {complete: false, id: 2, name: 'test2'}],
    loadingTrue: () => {},
    loadingFalse: () => {}
  }
    const tree = renderer
      .create(<MemoryRouter><GoalsContext.Provider value={state}><GoalDetail match={
        {params: {goalId: 1}}
      }/></GoalsContext.Provider></MemoryRouter>)
      .toJSON();
    expect(tree).toMatchSnapshot();  
  });
});