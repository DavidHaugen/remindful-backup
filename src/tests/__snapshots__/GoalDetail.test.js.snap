// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GoalDetail /> renders the UI as expected 1`] = `
<div
  className="main"
  role="main"
>
  <div
    className="wrapper"
  >
    <div
      role="alert"
    />
    <div
      className="goalTitle"
    >
      <h1
        className="inProgress overflowEl"
      >
        test
      </h1>
      <div
        className="goalButtons"
      >
        <div
          className="checkContainer"
          onClick={[Function]}
        >
          <i
            className="fas fa-check-circle"
          />
        </div>
        <div
          className="trashContainer"
          onClick={[Function]}
        >
          <i
            className="fas fa-trash"
          />
        </div>
      </div>
    </div>
    <div
      className="formContainer"
    >
      <form
        className="inputForm"
        id="reflectionsForm"
        onSubmit={[Function]}
      >
        <div
          role="alert"
        />
        <div
          className="formField"
        >
          <label
            className="inputLabel"
            htmlFor="reflectionsInput"
            id="addReflectionLabel"
          >
            Reflection
          </label>
          <input
            className="inputField"
            id="reflectionsInput"
            maxLength="55"
            name="reflectionsInput"
            required={true}
          />
          <button
            type="submit"
          >
            <i
              className="fas fa-plus-circle"
              id="reflectionButton"
              type="button"
            />
          </button>
        </div>
      </form>
    </div>
    <ul
      className="listContainer"
    >
      <div
        className="empty"
      >
        <p>
          No reflections yet.
        </p>
        Get started by adding a reflection above!
      </div>
    </ul>
  </div>
</div>
`;
